### ISSUE TITLE
Bossplot error message "inconsistent arcs"

### ISSUE THREAD

**Issue by kgage** [dkirkby/bossdata#74]
When attempting to plot using

```
$ bossplot --plate 3745 --mjd 55234 --fiber 882 --verbose
Plotting combined spectrum for 3745-55234-0882.
Found inconsistent arcs for expid 00107686.
```

What does this error message mean? The BOSS website has a spectrum, why am I not getting one?

**Comment by dkirkby** [comment]
This message is based on the headers `EXPIDnn` in HDU0.  For this plate, fibers 501-1000 give:

```
b2-00107686-00107690-00107685
b2-00107687-00107690-00107685
b2-00107688-00107690-00107685
b2-00107689-00107690-00107685
b2-00108116-00108120-00108121
b2-00108117-00108120-00108121
b2-00108118-00108120-00108121
b2-00108119-00108120-00108121
r2-00107686-00107690-00107691
r2-00107687-00107690-00107691
r2-00107688-00107690-00107691
r2-00107689-00107690-00107691
r2-00108116-00108120-00108121
r2-00108117-00108120-00108121
r2-00108118-00108120-00108121
r2-00108119-00108120-00108121
```

The three numbers are the science / flat / arc exposures.  The code expects to find matching pairs of `b2` and `r2` entries (or `b1` and `r1` for fibers 1-500) but, in this case, it appears that the red and blue cameras have _different_ arc exposures (00107685 vs 00107691) associated with the same science (and flat) exposure.

To debug this, look at the [spPlan2d file](http://dr12.sdss3.org/sas/dr12/boss/spectro/redux/v5_7_0/3745/spPlan2d-3745-55212.par) that was used to combine science exposures 00107686-00107689:

```
SPEXP 3745 55212 3745-55210-01 flat 30.0000 { sdR-b1-00107684.fit sdR-b2-00107684.fit sdR-r1-00107684.fit sdR-r2-00107684.fit }
SPEXP 3745 55212 3745-55210-01 arc 5.00000 { sdR-b1-00107685.fit sdR-b2-00107685.fit sdR-r1-00107685.fit sdR-r2-00107685.fit }
SPEXP 3745 55212 3745-55210-01 science 901.000 { sdR-b1-00107686.fit sdR-b2-00107686.fit sdR-r1-00107686.fit sdR-r2-00107686.fit }
SPEXP 3745 55212 3745-55210-01 science 900.000 { sdR-b1-00107687.fit sdR-b2-00107687.fit sdR-r1-00107687.fit sdR-r2-00107687.fit }
SPEXP 3745 55212 3745-55210-01 science 901.000 { sdR-b1-00107688.fit sdR-b2-00107688.fit sdR-r1-00107688.fit sdR-r2-00107688.fit }
SPEXP 3745 55212 3745-55210-01 science 901.000 { sdR-b1-00107689.fit sdR-b2-00107689.fit sdR-r1-00107689.fit sdR-r2-00107689.fit }
SPEXP 3745 55212 3745-55210-01 flat 30.0000 { sdR-b1-00107690.fit sdR-b2-00107690.fit sdR-r1-00107690.fit sdR-r2-00107690.fit }
SPEXP 3745 55212 3745-55210-01 arc 4.00000 { sdR-b1-00107691.fit sdR-b2-00107691.fit sdR-r1-00107691.fit sdR-r2-00107691.fit }
```

We see that this sequence begins and ends with a flat + arc calibration, and the headers consistently use the final flat (00107690) but use the initial arc (00107685) for the blue camera and the final arc (00107691) for the red camera.  Is this an error in the headers, an error in the pipeline, or the expected behavior?

The [log file](http://dr12.sdss3.org/sas/dr12/boss/spectro/redux/v5_7_0/3745/spDiag2d-3745-55212.log) contains the following clues:

```
SPREDUCE: b2: Best arc = sdR-b2-00107685.fit
SPREDUCE: r2: Best arc = sdR-r2-00107691.fit
```

Conclusions: the use of different arcs for the red/blue cameras is deliberate and `spec.get_exposure_table` should silently accept this.

**Comment by dkirkby** [comment]
The fix for this is a bit more involved that just removing a check since the current `exposure_table` data structure is based on a now invalid assumption.  Before (using fiber 1 to avoid this issue):

```
% bossplot --plate 3745 --mjd 55234 --fiber 1 --verbose
Plotting combined spectrum for 3745-55234-0001.
Exposure summary:
 exp    arc    flat   b1  b2  r1  r2
------ ------ ------ --- --- --- ---
107686 107691 107690   1   0   9   0
107687 107691 107690   2   0  10   0
107688 107691 107690   3   0  11   0
107689 107691 107690   4   0  12   0
108116 108121 108120   5   0  13   0
108117 108121 108120   6   0  14   0
108118 108121 108120   7   0  15   0
108119 108121 108120   8   0  16   0
```

**Comment by dkirkby** [comment]
After replacing the original `exposure_table` with a new `spec.Exposures` object:

```
% bossplot --plate 3745 --mjd 55234 --fiber 1 --verbose
Plotting combined spectrum for 3745-55234-0001.
Exposure summary:
offset camera science  flat   arc  
------ ------ ------- ------ ------
     0     b1  107686 107690 107691
     1     b1  107687 107690 107691
     2     b1  107688 107690 107691
     3     b1  107689 107690 107691
     4     b1  108116 108120 108121
     5     b1  108117 108120 108121
     6     b1  108118 108120 108121
     7     b1  108119 108120 108121
     8     r1  107686 107690 107691
     9     r1  107687 107690 107691
    10     r1  107688 107690 107691
    11     r1  107689 107690 107691
    12     r1  108116 108120 108121
    13     r1  108117 108120 108121
    14     r1  108118 108120 108121
    15     r1  108119 108120 108121
```

**Comment by dkirkby** [comment]
Testing with the spectrum that failed before now gives:

```
% bossplot --plate 3745 --mjd 55234 --fiber 882 --verbose
Plotting combined spectrum for 3745-55234-0882.
Exposure summary:
offset camera science  flat   arc  
------ ------ ------- ------ ------
     0     b2  107686 107690 107685
     1     b2  107687 107690 107685
     2     b2  107688 107690 107685
     3     b2  107689 107690 107685
     4     b2  108116 108120 108121
     5     b2  108117 108120 108121
     6     b2  108118 108120 108121
     7     b2  108119 108120 108121
     8     r2  107686 107690 107691
     9     r2  107687 107690 107691
    10     r2  107688 107690 107691
    11     r2  107689 107690 107691
    12     r2  108116 108120 108121
    13     r2  108117 108120 108121
    14     r2  108118 108120 108121
    15     r2  108119 108120 108121
```

and produces the plot below.  Note how the arc exposures for the same b2 and r2 science exposures are different, which was the root of the problem here, but that this is now handled sensibly. I am declaring victory!

![3745-55234-882](https://cloud.githubusercontent.com/assets/185007/8585233/0f102796-2594-11e5-9379-d483ee4b434a.png)

**Comment by dkirkby** [comment]
Fixed with PR #82
